File: cpp/llm/llama2_7b/embedding.pl
#include "ppl.h"

using namespace ppl;
__KERNEL__ void embedding_kernel(fp16 *ptr_output, fp16 *ptr_param,
                                 uint32 *ptr_index, int outer_num,
                                 int inner_num, int select_num, int index_num,
                                 int const_val) {
  set_core_num(8);
  int core_num = get_core_num();
  int core_idx = get_core_index();
  assert(outer_num == 1);
  if (select_num < inner_num) {
    int inner_slice = div_up(inner_num, core_num);
    int real_inner_slice = min(inner_slice, inner_num - core_idx * inner_slice);

    if (real_inner_slice <= 0) {
      return;
    }
    dim4 output_shape = {1, 1, index_num, inner_num};
    dim4 output_slice_shape = {1, 1, index_num, real_inner_slice};
    dim4 param_shape = {1, 1, select_num, inner_num};
    dim4 param_slice_shape = {1, 1, select_num, real_inner_slice};
    dim4 index_shape = {1, 1, index_num, 1};
    dim4 offset = {0, 0, 0, core_idx * inner_slice};

    auto output_g = gtensor<fp16>(output_shape, GLOBAL, ptr_output);
    auto param_g = gtensor<fp16>(param_shape, GLOBAL, ptr_param);
    auto index_g = gtensor<uint32>(index_shape, GLOBAL, ptr_index);

    dma::gather_h(output_g.sub_view(output_slice_shape, offset),
                  param_g.sub_view(param_slice_shape, offset), index_g, 0);

  } else {
    int index_slice = div_up(index_num, core_num);
    int allocated_core = div_up(index_num, index_slice);
    int real_index_slice = min(index_slice, index_num - core_idx * index_slice);

    if (core_idx >= allocated_core) {
      return;
    }
    dim4 output_shape = {1, 1, index_num, inner_num};
    dim4 output_slice_shape = {1, 1, real_index_slice, inner_num};
    dim4 param_shape = {1, 1, select_num, inner_num};
    dim4 index_shape = {1, 1, index_num, 1};
    dim4 index_slice_shape = {1, 1, real_index_slice, 1};
    dim4 offset = {0, 0, core_idx * index_slice, 0};

    auto output_g = gtensor<fp16>(output_shape, GLOBAL, ptr_output);
    auto param_g = gtensor<fp16>(param_shape, GLOBAL, ptr_param);
    auto index_g = gtensor<uint32>(index_shape, GLOBAL, ptr_index);

    dma::gather_h(output_g.sub_view(output_slice_shape, offset), param_g,
                  index_g.sub_view(index_slice_shape, offset), 0);
  }
}

__TEST__ void gather_h_s2s_index_global() {
#if 0
  int outer_num = 1;
  int select_num = 32000;
  int inner_num = 4096;
  int index_num = 6;
#else
  int outer_num = 1;
  int select_num = 15;
  int inner_num = 64;
  int index_num = 6;
#endif
  dim4 output_shape = {1, outer_num, index_num, inner_num};
  auto output = ppl::malloc<fp16>(&output_shape);

  dim4 param_shape = {1, outer_num, select_num, inner_num};
  auto param = ppl::malloc<fp16>(&param_shape);
  ppl::rand(param, &param_shape, -32, 32);

  dim4 index_shape = {1, outer_num, index_num, 1};
  auto index = ppl::malloc<uint32>(&index_shape);
  ppl::rand(index, &index_shape, 0, select_num -1);

  embedding_kernel(output, param, index, outer_num, inner_num, select_num,
                   index_num, 0);
}

File: python/llm/llama2_7b/embedding.py
import torch

import ppl
import ppl.language as pl
import struct
import numpy as np


@ppl.jit(debug=False)
def embedding_kernel(ptr_out, ptr_param, ptr_index, outer_num,
                     inner_num: pl.constexpr, select_num: pl.constexpr,
                     index_num):
    """
    embedding计算
        .. code-block:: python

             embedding_kernel(ptr_out, ptr_param, ptr_index, outer_num, inner_num, select_num, index_num)

    参数:
        - ``ptr_out`` (`ppl.language.tensor`): output张量

        - ``ptr_param`` (`ppl.language.tensor`): param张量

        - ``ptr_index`` (`ppl.language.tensor`): index张量

        - ``outer_num`` (`int`): outer_num

        - ``inner_num`` (`int`): inner_num

        - ``select_num`` (`int`): select_num

        - ``index_num`` (`int`): index_num

    返回值:
        无

    注意事项:
        无
    """
    core_idx = pl.get_core_index()
    core_num = pl.get_core_num()

    output_shape = [1, 1, index_num, inner_num]
    param_shape = [1, 1, select_num, inner_num]
    index_shape = [1, 1, index_num, 1]

    ptr_index.set_dtype(pl.pu32_t)

    output_g = pl.gtensor(output_shape, pl.GLOBAL, ptr_out)
    param_g = pl.gtensor(param_shape, pl.GLOBAL, ptr_param)
    index_g = pl.gtensor(index_shape, pl.GLOBAL, ptr_index)

    if (select_num < inner_num):
        inner_slice = pl.cast((inner_num + core_num - 1) // core_num, pl.int32)
        real_inner_slice = min(inner_slice, inner_num - core_idx * inner_slice)
        if real_inner_slice > 0:
            output_slice_shape = [1, 1, index_num, real_inner_slice]
            param_slice_shape = [1, 1, select_num, real_inner_slice]
            offset = [0, 0, 0, core_idx * inner_slice]
            pl.dma.gather_h(output_g.sub_view(output_slice_shape, offset),
                            param_g.sub_view(param_slice_shape, offset),
                            index_g, 0)
    else:
        index_slice = pl.cast((index_num + core_num - 1) // core_num, pl.int32)
        allocated_core = pl.cast((index_num + index_slice - 1) // index_slice,
                                 pl.int32)
        real_index_slice = min(index_slice, index_num - core_idx * index_slice)

        if core_idx < allocated_core:
            output_slice_shape = [1, 1, real_index_slice, inner_num]
            index_slice_shape = [1, 1, real_index_slice, 1]
            offset = [0, 0, core_idx * index_slice, 0]

            pl.dma.gather_h(output_g.sub_view(output_slice_shape,
                                              offset), param_g,
                            index_g.sub_view(index_slice_shape, offset))


if __name__ == '__main__':
    outer_num = 1
    select_num = 1024
    inner_num = 128
    index_num = 6

    param = torch.rand((select_num, inner_num), dtype=torch.float)
    index = torch.randint(0, 10, (1, index_num), dtype=torch.int32)
    index_cpu = index.to(torch.int32)
    output_ppl = torch.zeros((index_num, inner_num), dtype=torch.float)

    embedding_kernel[(
        1,
        8,
    )](output_ppl, param, index, outer_num, inner_num, select_num, index_num)

    output_torch = torch.nn.functional.embedding(index_cpu, param)

    atol = 1e-3  # 绝对容忍度
    rtol = 1e-3  # 相对容忍度
    # print(output_torch)
    # print(output_ppl)
    assert torch.allclose(output_torch, output_ppl, atol=atol,
                          rtol=rtol), "result cmp failed"
